lib/pwnlib/abi.rb

```diff
@@ -0,0 +1,97 @@
+# encoding: ASCII-8BIT
+
+require 'pwnlib/context'
+
+module Pwnlib
+  # Encapsulates information about a calling convention.
+  module ABI
+    # A super class for recording registers and stack's information.
+    class ABI
+      attr_reader :register_arguments
+      def initialize(regs, align, minimum)
+        @register_arguments = regs
+        @arg_alignment      = align
+        @stack_minimum      = minimum
+      end
+
+      def returns
+        true
+      end
+
+      def self.default
+        {
+          [32, 'i386', 'linux'] => LINUX_I386,
+          [64, 'amd64', 'linux'] => LINUX_AMD64,
+          [32, 'arm', 'linux'] => LINUX_ARM,
+          [32, 'thumb', 'linux'] => LINUX_ARM,
+          [32, 'mips', 'linux'] => LINUX_MIPS,
+          [32, 'i386', 'windows'] => WINDOWS_I386,
+          [64, 'amd64', 'windows'] => WINDOWS_AMD64
+        }[[context.bits, context.arch, context.os]]
+      end
+
+      def self.syscall
+        {
+          [32, 'i386', 'linux'] => LINUX_I386_SYSCALL,
+          [64, 'amd64', 'linux'] => LINUX_AMD64_SYSCALL,
+          [32, 'arm', 'linux'] => LINUX_ARM_SYSCALL,
+          [32, 'thumb', 'linux'] => LINUX_ARM_SYSCALL,
+          [32, 'mips', 'linux'] => LINUX_MIPS_SYSCALL
+        }[[context.bits, context.arch, context.os]]
+      end
+
+      def self.sigreturn
+        {
+          [32, 'i386', 'linux'] => LINUX_I386_SIGRETURN,
+          [64, 'amd64', 'linux'] => LINUX_AMD64_SIGRETURN,
+          [32, 'arm', 'linux'] => LINUX_ARM_SIGRETURN,
+          [32, 'thumb', 'linux'] => LINUX_ARM_SIGRETURN
+        }[[context.bits, context.arch, context.os]]
+      end
+      extend ::Pwnlib::Context
+    end
+
+    # The syscall ABI treats the syscall number as the zeroth argument,
+    # which must be loaded into the specified register.
+    class SyscallABI < ABI
+      def initialize(regs, align, minimum)
+        super
+        @syscall_register = regs[0]
+      end
+    end
+
+    # The sigreturn ABI is similar to the syscall ABI, except that
+    # both PC and SP are loaded from the stack.  Because of this, there
+    # is no 'return' slot necessary on the stack.
+    class SigreturnABI < SyscallABI
+      def returns
+        false
+      end
+    end
+    LINUX_I386    = ABI.new([], 4, 0)
+    LINUX_AMD64   = ABI.new(%w(rdi rsi rdx rcx r8 r9), 8, 0)
+    LINUX_ARM     = ABI.new(%w(r0 r1 r2 r3), 8, 0)
+    LINUX_AARCH64 = ABI.new(%w(x0 x1 x2 x3), 16, 0)
+    LINUX_MIPS    = ABI.new(%w($a0 $a1 $a2 $a3), 4, 0)
```

Comment:
Should we just don't do `=` alignment at all? Seems current code has some aligned and some not.
I prefer not.